File: trustkit/src/main/java/com/datatheorem/android/trustkit/utils/VendorIdentifier.java
package com.datatheorem.android.trustkit.utils;


import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import java.util.UUID;

/**
 * When TrustKit sends a report, it also sends a randomly-generated identifier to uniquely identify
 * a specific App install (ie. an instance of the App running on a specific device). It is the least
 * intrusive way to detect reports coming from the same device.
 */
public class VendorIdentifier {

    private static final String TRUSTKIT_VENDOR_ID = "TRUSTKIT_VENDOR_ID";

    @NonNull
    public static String getOrCreate(@NonNull Context appContext) {
        SharedPreferences trustKitSharedPreferences =
                PreferenceManager.getDefaultSharedPreferences(appContext);
        // We store the vendor ID in the App's preferences
        String appVendorId = trustKitSharedPreferences.getString(TRUSTKIT_VENDOR_ID, "");
        if (appVendorId.equals("")) {
            // First time the App is running: generate and store a new vendor ID
            TrustKitLog.i("Generating new vendor identifier...");
            appVendorId = UUID.randomUUID().toString();
            SharedPreferences.Editor editor = trustKitSharedPreferences.edit();
            editor.putString(TRUSTKIT_VENDOR_ID, appVendorId);
            editor.apply();
        }
        return appVendorId;
    }
}

File: trustkit/src/main/java/com/datatheorem/android/trustkit/utils/VendorIdentifier.kt
@file:JvmName("VendorIdentifier")

package com.datatheorem.android.trustkit.utils

import android.content.Context
import android.content.Context.MODE_PRIVATE
import android.preference.PreferenceManager
import java.util.UUID

/**
 * When TrustKit sends a report, it also sends a randomly-generated identifier to uniquely identify
 * a specific App install (ie. an instance of the App running on a specific device). It is the least
 * intrusive way to detect reports coming from the same device.
 */
private const val TRUSTKIT_VENDOR_ID = "TRUSTKIT_VENDOR_ID"

fun getOrCreate(appContext: Context): String {
    // new sdk preferences
    val tkSharedPreferences =
        appContext.getSharedPreferences("DataTheoremPreferences", MODE_PRIVATE)

    val appVendorId = tkSharedPreferences.getString(TRUSTKIT_VENDOR_ID, "") ?: ""
    if (appVendorId.isNotBlank()) return appVendorId

    // legacy default preferences
    val defaultPreferences = PreferenceManager.getDefaultSharedPreferences(appContext)

    val defaultAppVendorId = defaultPreferences.getString(TRUSTKIT_VENDOR_ID, "") ?: ""
    // we have a legacy id, write it to the new preferences before returning
    if (defaultAppVendorId.isNotBlank()) {
        tkSharedPreferences.edit().putString(TRUSTKIT_VENDOR_ID, defaultAppVendorId).apply()
        return defaultAppVendorId
    }

    // we don't have an id, generate one
    TrustKitLog.i("Generating new vendor identifier...")
    val newId = UUID.randomUUID().toString()
    tkSharedPreferences.edit().putString(TRUSTKIT_VENDOR_ID, newId).apply()

    return newId
}
